File: commands/stats.js
import { SlashCommandBuilder, EmbedBuilder, MessageFlags } from 'discord.js';
import { readTournamentData, readTournamentConfig, readPlayers, readAllVotes, readPlayerBadges } from '../utils/firebase.js';
import { formatBadgesDetailed } from '../utils/badges.js';
import { getWinner, getMatchVote, VND_FORMATTER } from '../utils/helper.js';
import logger from '../utils/logger.js';

export const data = new SlashCommandBuilder()
  .setName('stats')
  .setDescription('Thống kê cá nhân — đối mặt sự thật đi')
  .addUserOption(option => option.setName('user')
    .setDescription('Soi ai (bỏ trống = soi mình)')
    .setRequired(false));

export async function execute(interaction) {
  try {
    const config = await readTournamentConfig();
    const tournamentName = config?.name || 'Tournament';
    const targetUser = interaction.options.get('user')?.user || interaction.user;
    const userId = targetUser.id;

    const players = (await readPlayers()).val();
    if (!players || !players[userId]) {
      const embed = new EmbedBuilder()
        .setTitle('❌  Chưa đăng ký')
        .setDescription(
          targetUser.id === interaction.user.id
            ? 'Chưa đăng ký kìa. `/register` đi rồi chơi!'
            : `${targetUser} chưa đăng ký.`
        )
        .setColor(0xED4245);
      await interaction.reply({ embeds: [embed], flags: MessageFlags.Ephemeral });
      return;
    }

    const player = players[userId];
    const allMatches = (await readTournamentData('matches')).val();
    const votes = await readAllVotes();

    let correctVotes = 0;
    let totalVotes = 0;
    const recentResults = [];

    const completedMatches = (allMatches || [])
      .filter((m) => m.hasResult && m.isCalculated)
      .sort((a, b) => Date.parse(a.date) - Date.parse(b.date));

    for (const match of completedMatches) {
      const key = `${match.id - 1}`;
      const userVote = getMatchVote(votes, key, match.messageId, userId);
      if (userVote === null) continue;

      const winner = getWinner(match);
      const isCorrect = userVote === winner;

      totalVotes++;
      if (isCorrect) correctVotes++;

      recentResults.push({
        matchId: match.id,
        home: match.home,
        away: match.away,
        vote: userVote,
        correct: isCorrect,
      });
    }

    const winRate = totalVotes > 0 ? ((correctVotes / totalVotes) * 100).toFixed(0) : '—';
    const nickname = interaction.client.cachedUsers[userId]?.nickname || targetUser.displayName;

    const embed = new EmbedBuilder()
      .setTitle(`📊  Thống kê ${nickname}`)
      .setDescription(`**${tournamentName}**`)
      .setColor(0x5865F2)
      .setThumbnail(targetUser.displayAvatarURL())
      .addFields(
        { name: '💰 Tài khoản', value: VND_FORMATTER.format(player.points * 1000), inline: true },
        { name: '🎮 Đã chơi', value: `${player.matches} trận`, inline: true },
        { name: '🎯 Tỉ lệ đúng', value: totalVotes > 0 ? `${winRate}% (${correctVotes}/${totalVotes})` : 'Chưa vote', inline: true },
      )
      .setTimestamp();

    const recent = recentResults.slice(-5).reverse();
    if (recent.length > 0) {
      const lines = recent.map((r) => {
        const icon = r.correct ? '👑' : '🤡';
        return `${icon} #${r.matchId} ${r.home.toUpperCase()} vs ${r.away.toUpperCase()} — vote **${r.vote.toUpperCase()}**`;
      });
      embed.addFields({ name: '🕐 Mấy trận gần đây', value: lines.join('\n'), inline: false });
    }

    const storedBadges = await readPlayerBadges(userId);
    embed.addFields({
      name: '🏅 Huy hiệu',
      value: formatBadgesDetailed(storedBadges),
      inline: false,
    });

    await interaction.reply({ embeds: [embed] });
  } catch (err) {
    logger.error(err);
    if (!interaction.replied && !interaction.deferred) {
      await interaction.reply({ content: '❌ Không thể tải thống kê.', flags: MessageFlags.Ephemeral }).catch(() => {});
    }
  }
}

File: commands/stats.go
package commands

import (
	"context"
	"fmt"
	"sort"
	"time"

	"github.com/kdnvi/rambo-bot/bot"
	"github.com/kdnvi/rambo-bot/internal/discord"
)

func Stats(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	// Resolve target user
	targetID := i.ActingUser().ID
	targetAvatarURL := i.ActingUser().AvatarURL()
	targetDisplayName := i.ActingUser().DisplayName()

	if opt := i.Data.GetOption("user"); opt != nil {
		targetID = opt.String()
	}

	players, err := b.DB.ReadPlayers(ctx)
	if err != nil || players[targetID] == nil {
		msg := "Chưa đăng ký kìa. `/register` đi rồi chơi!"
		if targetID != i.ActingUser().ID {
			msg = fmt.Sprintf("<@%s> chưa đăng ký.", targetID)
		}
		replyEphemeral(ctx, b, i, discord.InteractionCallbackData{
			Embeds: []discord.Embed{{
				Title:       "❌  Chưa đăng ký",
				Description: msg,
				Color:       0xED4245,
			}},
		})
		return
	}

	if err := deferReply(ctx, b, i, false); err != nil {
		return
	}

	tournamentName := getTournamentName(ctx, b)
	player := players[targetID]

	matches, _ := b.DB.ReadMatches(ctx)
	votes, _ := b.DB.ReadAllVotes(ctx)
	userWagers, _ := b.DB.ReadUserWagers(ctx, targetID)

	completed := make([]int, 0)
	for idx, m := range matches {
		if m.HasResult && m.IsCalculated {
			completed = append(completed, idx)
		}
	}
	sort.Slice(completed, func(a, b int) bool {
		ta, _ := time.Parse(time.RFC3339, matches[completed[a]].Date)
		tb, _ := time.Parse(time.RFC3339, matches[completed[b]].Date)
		return ta.Before(tb)
	})

	correctCount, votedCount := 0, 0
	type recentEntry struct {
		matchID int
		home    string
		away    string
		vote    string
		correct bool
		auto    bool
		random  bool
	}
	var recentResults []recentEntry

	for _, idx := range completed {
		m := matches[idx]
		key := fmt.Sprintf("%d", m.ID-1)
		userVote := getMatchVote(votes, key, m.MessageID, targetID)
		winner := getWinner(&m)

		voted := userVote != ""
		correct := voted && userVote == winner
		if voted {
			votedCount++
		}
		if correct {
			correctCount++
		}

		wager := userWagers[fmt.Sprintf("%d", m.ID)]
		recentResults = append(recentResults, recentEntry{
			matchID: m.ID,
			home:    m.Home,
			away:    m.Away,
			vote:    userVote,
			correct: correct,
			auto:    !voted,
			random:  wager != nil && wager.Random,
		})
	}

	winRate := "—"
	if votedCount > 0 {
		winRate = fmt.Sprintf("%d%% (%d/%d)", (correctCount*100)/votedCount, correctCount, votedCount)
	}

	// Resolve display name from cache if available
	if u := b.GetCachedUser(targetID); u != nil {
		targetDisplayName = u.DisplayName()
		targetAvatarURL = u.AvatarURL
	}

	embed := discord.Embed{
		Title:       fmt.Sprintf("📊  Thống kê %s", targetDisplayName),
		Description: fmt.Sprintf("**%s**", tournamentName),
		Color:       0x5865F2,
		Fields: []discord.EmbedField{
			{Name: "💰 Tài khoản", Value: fmtVND(player.Points), Inline: true},
			{Name: "🎮 Đã chơi", Value: fmt.Sprintf("%d trận", player.Matches), Inline: true},
			{Name: "🎯 Tỉ lệ đúng", Value: winRate, Inline: true},
		},
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
	if targetAvatarURL != "" {
		embed.Thumbnail = &discord.EmbedImage{URL: targetAvatarURL}
	}

	// Last 5 matches
	recent := recentResults
	if len(recent) > 5 {
		recent = recent[len(recent)-5:]
	}
	// reverse
	for lo, hi := 0, len(recent)-1; lo < hi; lo, hi = lo+1, hi-1 {
		recent[lo], recent[hi] = recent[hi], recent[lo]
	}
	if len(recent) > 0 {
		lines := make([]string, 0, len(recent))
		for _, r := range recent {
			icon := "🤡"
			if r.correct {
				icon = "👑"
			}
			voteLabel := fmt.Sprintf("vote **%s**", upper(r.vote))
			if r.auto {
				if r.random {
					voteLabel = "🎲 random"
				} else {
					voteLabel = "🤖 auto"
				}
			}
			lines = append(lines, fmt.Sprintf("%s #%d %s vs %s — %s",
				icon, r.matchID, upper(r.home), upper(r.away), voteLabel))
		}
		embed.Fields = append(embed.Fields, discord.EmbedField{
			Name:  "🕐 Mấy trận gần đây",
			Value: joinLines(lines, "\n"),
		})
	}

	// Badges
	storedBadges, _ := b.DB.ReadPlayerBadges(ctx, targetID)
	embed.Fields = append(embed.Fields, discord.EmbedField{
		Name:  "🏅 Huy hiệu",
		Value: formatBadgesDetailed(storedBadges),
	})

	editReply(ctx, b, i, discord.EditMessagePayload{Embeds: []discord.Embed{embed}})
}
